### PR TITLE
go-server: Update server to use Golang

### PR DESCRIPTION
**Pull request by zonggen**
Use Golang as back-end language, ~still WIP..~

Signed-off-by: Allen Bai <abai@redhat.com>

### CODE REVIEW

**Comment by zonggen** [comment]
STATUS:
 - [x] Migrate `app.py` to `main.go`
 - [x] Update `Dockerfile` and stop using gunicorn
 - [x] Add test cases for `main.go`
 - [x] Update `Makefile`

File: server/main.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+// configHandler handles config requests when they come as a POST
+// note: to enable CORS, it should handle OPTIONS as well
+func configHandler(w http.ResponseWriter, r *http.Request) {
+	// enable CORS
+	w.Header().Set("Access-Control-Allow-Origin", "*")
+	w.Header().Set("Access-Control-Allow-Methods", "POST, OPTIONS")
+	w.Header().Set("Access-Control-Allow-Headers", "Accept, Content-Type, Content-Length, Accept-Encoding, X-CSRF-Token, Authorization")
+
+	if r.Method == "OPTIONS" {
```
Comment (ashcrow) [inline server/main.go:84]: I'd recommend moving `// note: to enable CORS, it should handle OPTIONS as well` to this line.

**Review by ashcrow — COMMENTED** [review]
Looks really good! Have you tested it?

File: server/main.go (hunk 1 above)
Comment (zonggen) [reply server/main.go:84]: Thanks for the quick review!
Updated

**Comment by zonggen** [comment]
@ashcrow Yep, tested and it should run without problem

File: server/main.go [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+	(*w).Header().Set("Content-Type", "application/json")
+	(*w).WriteHeader(status)
+	(*w).Write(resJSON)
+}
+
+// runCommand runs the command and returns the error, stdout, and stderr
+func runCommand(command string, input string) (string, string, error) {
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+	cmd := exec.Command("bash", "-c", command)
```
Comment (ashcrow) [inline server/main.go]: Why run with bash? It's possible that the command could be crafted to inject scripting.

Comment (zonggen) [reply server/main.go]: Sounds not secure, so in this case just `exe.Command("./fcct", "input_config")` would be better?

**Comment by zonggen** [comment]
This PR doesn't seem to trigger new build on Travis CI anymore...

File: server/main.go (hunk 2 above)
Comment (ashcrow) [reply server/main.go]: I believe it would be better.

Comment (zonggen) [reply server/main.go]: Updated

**Comment by zonggen** [comment]
Built on quay.io,
`podman run --rm -d --name fcct-online -e "PORT=8765" -p 8007:8765 fcct-online:migrate-to-go` 
should now run on `http://localhost:8007`

**Comment by ashcrow** [comment]
Checking now @zonggen

File: server/main.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+		}
+
+		// checks if `config_string` has reasonal length
+		maxLen := 31415
+		maxLenStr := os.Getenv("ONLINE_FCCT_MAX_LENGTH")
+		if len(maxLenStr) > 0 {
+			maxLen, _ = strconv.Atoi(maxLenStr)
+		}
+		if len(pd.ConfigString) > maxLen {
+			res.ErrMsg = "failed: FCC string too long"
```
Comment (ashcrow) [inline server/main.go:129]: This doesn't end up going back to the client for some reason, but it can be fixed later.

File: server/main.go [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+		// checks if POST json contains a `config_string` key
+		if pd.ConfigString == "default" {
+			res.ErrMsg = "failed: config_string must be set"
+			log.Printf(res.ErrMsg)
+			writeResponse(&w, res, http.StatusBadRequest)
+			return
+		}
+
+		// checks if `config_string` has reasonal length
+		maxLen := 31415
```
Comment (ashcrow) [inline server/main.go:123]: In the future (not now) this should probably be set at run time instead of reset on every call.

File: server/main.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+		// otherwise FCCT exits with 0
+		// and the stdout has correct FCC
+		log.Printf(out)
+		log.Printf(stderr)
+
+		var messageJSON map[string]interface{}
+		_ = json.Unmarshal([]byte(out), &messageJSON)
+
+		log.Println("Successfully convert to Ignition config")
```
Comment (ashcrow) [inline server/main.go:167]: This log line as is probably isn't needed.

File: server/main.go [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+// configHandler handles config requests when they come as a POST
+func configHandler(w http.ResponseWriter, r *http.Request) {
+	// enable CORS
+	w.Header().Set("Access-Control-Allow-Origin", "*")
+	w.Header().Set("Access-Control-Allow-Methods", "POST, OPTIONS")
+	w.Header().Set("Access-Control-Allow-Headers", "Accept, Content-Type, Content-Length, Accept-Encoding, X-CSRF-Token, Authorization")
+
+	// note: to enable CORS, it should handle OPTIONS as well
+	if r.Method == "OPTIONS" {
```
Comment (ashcrow) [inline server/main.go:84]: Looks like OPTIONS ends up being blocked by nginx, but it's not causing a problem.

```
$ curl -X OPTIONS http://127.0.0.1:8007
<html>
<head><title>405 Not Allowed</title></head>
<body bgcolor="white">
<center><h1>405 Not Allowed</h1></center>
<hr><center>nginx/1.14.1</center>
</body>
</html>
```

Comment (ashcrow) [reply server/main.go:84]: Scratch that, it works for `/config`

**Review by ashcrow — APPROVED** [review]
Tests pass, functionality works, I feel this is good to merge. Any nits/improvements/fixes can be iterated on later as needed. Good work!

File: server/main.go (hunk 3 above)
Comment (zonggen) [reply server/main.go:129]: I tried with `ONLINE_FCCT_MAX_LENGTH=10 ./fcct-online --debug` and looks like it works..

**Pull request merged**